### ISSUE TITLE
OpenBLAS is not portable across different CPU architectures

### ISSUE THREAD

**Issue by mlell** [nickjer/singularity-r#6]
Hi,

you have merged my pull request #3, that uses OpenBLAS. As we are currently experiencing, this might be a problem when moving the container between machines, as OpenBLAS is specialized for different CPU architectures at compile-time. I do not know yet how this is handled in the corresponding Debian package, I just want to let you know as soon as possible.

Cheers, 
Moritz

**Comment by nickjer** [comment]
Are you using my deployed version at `library://nickjer/default/singularity-r:4.1.1`? That was built using my local machine with the `x86_64` architecture. I was hoping the debian package would be able to work across all CPUs that are `x86_64` irrespective of instruction sets it can support (e.g., SSE, SIMD, ...). Is this what you are not experiencing?

I can imagine this will not work on `i686` or `arm` CPU architectures. So I may need to leave a note in the [README.md](https://github.com/nickjer/singularity-r/blob/master/README.md) to only  use the pre-built image for `x86_64`, all others need to rebuild the image on the respective machine.

**Comment by mlell** [comment]
At our site, we saw different results for `tcrossprod(X)` (only with one argument) and `lqmm::is.positive.definite()`. The problem is correlated with the processor microarchitecture (`cat /sys/devices/cpu/caps/pmu_name`). On the one side are "westmere", "nehalem", "haswell" and "ivybridge", on the other side is "skylake". 

The image we use is mlell/singularity-r, which is derived from this repo but heavily altered to include user-friendly start/stop scripts and more libraries. But for OpenBLAS the debian package libopenblas-dev is used, which should be the same as here, though I am not sure about the version. 

In the [description of the debian package](https://packages.debian.org/de/sid/libopenblas-dev) they say 

> On amd64, arm64, i386, ppc64el, s390x, kfreebsd-amd64 and kfreebsd-i386, all kernels are included in the library and the one matching best your processor is selected at runtime.

So I was hoping the same as you. I am currently tracking this further down, I just do not know how long it takes and wanted to post the suspicion early because there is no error, just different results which can be quite a pernicious bug in analyses.

**Comment by nickjer** [comment]
Thanks for following up on this. In the meantime I can update the `README.md` to reflect that building the image on the machine you intend to run it on is best for OpenBLAS.

**Comment by mlell** [comment]
There is also an [old discussion at the OpenBLAS repo](https://github.com/xianyi/OpenBLAS/issues/685) how to build a version that is supposed to be more portable. Not sure though whether the mentioned GENERIC target is still relevant, I do not find it in the OpenBLAS `TargetsList.txt` file.

**Comment by mlell** [comment]
Reported to OpenBLAS (xianyi/OpenBLAS#3454), which seems to be a duplicate of xianyi/OpenBLAS#2986. Following that, the issue is not determined by the machine where the Singularity image is built, but rather by the machine where the debian package is built. So building the Singularity image from scratch as it is currently written in the README.md might not improve things because the OpenBLAS binaries are not rebuilt by that process. To solve the problem, OpenBLAS must be recompiled during the image building process.

**Comment by mlell** [comment]
Currently, we solve that by not including OpenBLAS in the image. To enable it, one can 

```
git clone https://github.com/xianyi/openblas
cd openblas
make -j
cd ..

singularity run -B "$(readlink -f "$PWD/openblas/libopenblas.so"):/usr/lib/R/lib/libblas.so.3" IMAGE.sif
```

The library name that R expects can be found by `ldd /usr/lib/R/bin/exec/R`. 
The usual way to include host libraries into the container is to bind them to "/.singularity.d/libs", but R alters `LD_LIBRARY_PATH` before starting (see `/usr/lib/R/etc/ldpaths`), so it needs to be injected directly into the R installation.

**Comment by nickjer** [comment]
Thanks, I updated the `README.md` with these instructions.
